File: ats/public/js/delivery_note_custom.js
frappe.ui.form.on('Delivery Note', {
    refresh: function(frm) {

        if (frm.doc.docstatus === 1) {
            frm.add_custom_button(
                __("Sales Invoice"),
                () => frm.events.map_to_sales_invoice(frm),
                __("Create")
            );
        }
    },

    map_to_sales_invoice: function(frm) {

        frappe.call({
            method: "ats.custom.delivery_note_custom.create_sales_invoice",
            args: {
                delivery_note: frm.doc.name
            },
            freeze: true,
            freeze_message: __("Creating Sales Invoice..."),
            callback: function(r) {
                if (r.message) {
                    const si_doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Sales Invoice", si_doc.name);
                }
            }
        });

    },
});

File: ats/custom/delivery_note_custom.py
import frappe
from frappe.model.mapper import get_mapped_doc

@frappe.whitelist()
def create_sales_invoice(delivery_note):
    doc = get_mapped_doc(
        "Delivery Note",
        delivery_note,
        {
            "Delivery Note": {
                "doctype": "Sales Invoice",
                "field_map": {
                    "customer": "customer",
                    "name": "delivery_note"   # optional reference
                }
            },
            "Delivery Note Item": {
                "doctype": "Sales Invoice Item",
                "field_map": {
                    "item_code": "item_code",
                    "item_name": "item_name",
                    "qty": "qty",
                    "uom": "uom",
                    "rate": "rate",
                    "amount": "amount"
                }
            }
        }
    )

    return doc
